core/src/main/scala/com/yahoo/maha/core/query/QueryGenerator.scala

```diff
@@ -260,18 +260,45 @@ object QueryGeneratorHelper {
   }
 }
 
+
+case class Version(value: Int)
+object Version {
+  val DEFAULT = V0
+  def of(ver: Int): Version = {
+    ver match {
+      case 0 => V0
+      case 1 => V1
+      case 2 => V2
+    }
+  }
+}
+object V0 extends Version(0)
+object V1 extends Version(1)
+object V2 extends Version(2)
+
 class QueryGeneratorRegistry {
-  private[this] var queryGeneratorRegistry : Map[Engine, QueryGenerator[_]] = Map.empty
 
-  def isEngineRegistered(engine: Engine): Boolean = synchronized {
-    queryGeneratorRegistry.contains(engine)
+  private[this] var queryGeneratorRegistry : Map[Engine, Map[Version , QueryGenerator[_]]] = Map.empty
+
+  def isEngineRegistered(engine: Engine, version: Version = Version.DEFAULT): Boolean = synchronized {
+    queryGeneratorRegistry.contains(engine) && queryGeneratorRegistry.get(engine).get.contains(version)
   }
 
-  def register[U <: QueryGenerator[_]](engine: Engine, qg: U) : Unit = synchronized {
-    require(!queryGeneratorRegistry.contains(engine), s"Query generator already defined for engine : $engine")
-    queryGeneratorRegistry += (engine -> qg)
+  def register[U <: QueryGenerator[_]](engine: Engine, qg: U, version: Version = Version.DEFAULT) : Unit = synchronized {
+    require(!isEngineRegistered(engine, version), s"Query generator already defined for engine : $engine and version $version")
+    if (queryGeneratorRegistry.contains(engine)) {
+      queryGeneratorRegistry += (engine -> (Map(version -> qg) ++ queryGeneratorRegistry.get(engine).get))
+    } else {
+      queryGeneratorRegistry += (engine -> Map(version -> qg))
+    }
+  }
+
+  def getGenerator(engine: Engine, version: Option[Version]): Option[QueryGenerator[EngineRequirement]] = {
+    if (queryGeneratorRegistry.get(engine).isDefined) {
```

Comment:
Could use for yield instead of if else block
```
for {
  versionGeneratorMap <- queryGeneratorRegistry.get(engine)
  generator <- versionGeneratorMap.get(version.getOrElse(Version.DEFAULT))
} yield generator